File: csoj-backend-master/src/main/java/com/oj/cs/service/DailyChallengeService.java
package com.oj.cs.service;

import java.time.LocalDate;
import java.util.List;

import com.baomidou.mybatisplus.extension.service.IService;
import com.oj.cs.model.entity.ChallengeStreak;
import com.oj.cs.model.entity.DailyChallenge;
import com.oj.cs.model.entity.UserChallenge;

/** 每日挑战系统服务 */
public interface DailyChallengeService extends IService<DailyChallenge> {

  /**
   * 获取今日挑战
   *
   * @return 今日挑战
   */
  DailyChallenge getTodayChallenge();

  /**
   * 获取指定日期的挑战
   *
   * @param date 日期
   * @return 挑战
   */
  DailyChallenge getChallengeByDate(LocalDate date);

  /**
   * 完成挑战
   *
   * @param userId 用户ID
   * @param challengeId 挑战ID
   * @param isPerfect 是否完美通关
   * @return 获得的积分
   */
  int completeChallenge(Long userId, Long challengeId, boolean isPerfect);

  /**
   * 获取用户挑战状态
   *
   * @param userId 用户ID
   * @param challengeId 挑战ID
   * @return 用户挑战记录
   */
  UserChallenge getUserChallengeStatus(Long userId, Long challengeId);

  /**
   * 获取用户连胜信息
   *
   * @param userId 用户ID
   * @return 连胜信息
   */
  ChallengeStreak getUserStreak(Long userId);

  /**
   * 获取挑战历史
   *
   * @param userId 用户ID
   * @param days 天数
   * @return 挑战历史列表
   */
  List<UserChallenge> getChallengeHistory(Long userId, int days);

  /**
   * 生成每日挑战(管理员或定时任务调用)
   *
   * @return 生成的挑战
   */
  DailyChallenge generateDailyChallenge();

  /**
   * 检查题目是否为今日挑战
   *
   * @param questionId 题目ID
   * @return 是否为今日挑战
   */
  boolean isTodayChallenge(Long questionId);

  /**
   * 处理题目提交事件
   *
   * @param userId 用户ID
   * @param questionId 题目ID
   * @param accepted 是否通过
   */
  void onQuestionSubmit(Long userId, Long questionId, boolean accepted);
}

File: csoj-frontend-master/src/services/DailyChallengeService.ts
import axios from "axios";

const API_BASE = "/api";

/**
 * 每日挑战服务
 */
class DailyChallengeService {
  /**
   * 获取今日挑战
   */
  async getTodayChallenge() {
    const response = await axios.get(`${API_BASE}/challenge/today`);
    return response.data;
  }

  /**
   * 完成挑战
   * @param challengeId 挑战ID
   * @param isPerfect 是否完美通关
   */
  async completeChallenge(challengeId: number, isPerfect: boolean = false) {
    const response = await axios.post(`${API_BASE}/challenge/complete`, null, {
      params: { challengeId, isPerfect },
    });
    return response.data;
  }

  /**
   * 获取用户挑战状态
   * @param challengeId 挑战ID
   */
  async getUserChallengeStatus(challengeId: number) {
    const response = await axios.get(`${API_BASE}/challenge/status/${challengeId}`);
    return response.data;
  }

  /**
   * 获取用户连胜信息
   */
  async getUserStreak() {
    const response = await axios.get(`${API_BASE}/challenge/streak`);
    return response.data;
  }

  /**
   * 获取挑战历史
   * @param days 天数
   */
  async getChallengeHistory(days: number = 30) {
    const response = await axios.get(`${API_BASE}/challenge/history`, {
      params: { days },
    });
    return response.data;
  }
}

export default new DailyChallengeService();
